Implement github repo ingestion

tarpdalton opened the pull request:
Creates github repo node in ground with a node version for each commit
Still need to pull in the latest code for looking up tags so that commits/versions can link.
This PR addresses: #9

tarpdalton committed:
project setup

tarpdalton committed:
update yml with ports

api is working locally

tarpdalton committed:
ignore files

tarpdalton committed:
model for githubwebhook

tarpdalton committed:
Make kafka producer accessible

push webhook to kakfa

tarpdalton committed:
Merge remote-tracking branch 'ground-context/master'

tarpdalton committed:
Merge remote-tracking branch 'ground-context/master'

tarpdalton committed:
code formatting

tarpdalton committed:
parser working

create config for git parser
dedup latest node output
still need to POST to ground

tarpdalton committed:
posting to ground

need a way to look up previous commits in ground. or else manage it in
the script

tarpdalton committed:
github model test passing

tarpdalton committed:
add kafkaparams in yaml

tarpdalton committed:
revert changes

tarpdalton committed:
fix dependencies

tarpdalton committed:
git ignore update

tarpdalton committed:
git history parser working

vsreekanti commented on ground-core/pom.xml:
```diff
@@ -182,6 +182,27 @@
+        <dependency>
+            <groupId>org.apache.kafka</groupId>
+            <artifactId>kafka-clients</artifactId>
+            <version>0.10.0.1</version>
```
Can we pull version numbers out and put them in the root `pom.xml` and use variables? That way, we can enforce version uniformity across the project.

vsreekanti reviewed (commented):
How come there's all this (what seems like) autogenerated code for these Github models? It seems like we shouldn't really need any of that, and I'm a little uncomfortable with adding a bunch of code to the model that's not really part of the "core" data model. Is there any reason that we can't take the JSON from the webhook and parse it and send it directly to the Python server?

tarpdalton committed:
Removed webhook model

paramaterized kafka version
changed to kafka resource instead of github webhook resource

tarpdalton committed:
docker build working

tarpdalton committed:
swagger API documentation implemented

#16

vsreekanti commented on plugins/github-plugin/parsegitlog.py:
```diff
@@ -0,0 +1,133 @@
+    sourceLog = sourceLog.split("\n")
+    sourceLog = [[row.split("\t")[0], row.split("\t")[1].split(" ")[0]] for row in sourceLog]
+
+    latests = requests.get(latestUrl).text  # get latest commits # TODO get the latest commit
```
Is this TODO still pending?

tarpdalton replied:
Yes I need to finish this logic before the PR is merged, or else it will add all commits(node versions) every time.

vsreekanti commented on plugins/github-plugin/parsegitlog.py:
```diff
@@ -0,0 +1,133 @@
+from kafka import KafkaConsumer
+import configparser
+
+# TODO manage the kafka offset for restarts, error handling
```
We should be able to just write this into a local file, right?

vsreekanti commented on plugins/github-plugin/parsegitlog.py:
```diff
@@ -0,0 +1,133 @@
+    nodeVUrl = url + '/nodes/versions/'
+    latestUrl = nodeUrl + '/latest'
+
+    parentsLog = g.log('--all', '--format="%H,%P"')  # log with commit hash and parent hashes
```
Which git library are you using? Does this work as a plumbing command?

tarpdalton replied:
I'm using [gitpython](https://github.com/gitpython-developers/GitPython) right now. The other option is to use `subprocess.run` to run commands like `git rev-list` and return the output. I can't find a way to get the original branch name without using `git log --source` though. Log isn't a plumbing command, but since the library handles it it returns a string.

vsreekanti commented on ground-core/src/main/java/edu/berkeley/ground/resources/KafkaResource.java:
```diff
@@ -0,0 +1,59 @@
+  public String sendKafkaMessage(@Valid String value, @ApiParam(value = "Topic to push to", required = true) @QueryParam("topic") String topic, @ApiParam(value = "Key of message", required = true) @QueryParam("key") String key) throws GroundException {
+      LOGGER.info("Receiving kafka publish");
+      KafkaProduce.push(kafkaHost, kafkaPort, topic, key, value);
+  return value;
```
formatting

vsreekanti commented on ground-core/pom.xml:
```diff
@@ -248,6 +265,23 @@
             </plugin>
+
+            <plugin>
+                <groupId>com.spotify</groupId>
```
What does this plugin do?

tarpdalton replied:
This builds the docker image if you run  
`mvn clean package docker:build`  
We could also add docker:build in one of the lifecycle phases. So that it automatically gets built and pushed. We can also remove the plugin and do it manually if we want fewer dependencies.
[docker maven plugin](https://github.com/spotify/docker-maven-plugin)

vsreekanti commented on ground-core/docker/config.yml:
```diff
@@ -0,0 +1,15 @@
+kafkaHost: localhost
+kafkaPort: 9092
+server:
+  applicationConnectors:
```
Any reaosn we should default to `9090` instead of `8080`?

tarpdalton replied:
No it should be 8080, I can't remember why I put 9090

Commit:
Add Coveralls integration and fix Neo4j on Travis

Commit:
change .travis.yml to install Neo4j manually

Commit:
Remove part of neo4j install that's not necessary

Commit:
Change scripts to only drop data if asked to

Commit:
change test scripts to explicitly drop tables

Commit:
Move off container based

Commit:
Install newer Cassandra manually

Commit:
get rid of default Cassandra service

Commit:
attempt to fix Cassandra start

tarpdalton committed:
fixed ports

docker quickstart

tarpdalton committed:
update python config

tarpdalton committed:
apache license

tarpdalton committed:
Merge remote-tracking branch 'ground-context/master'

# Conflicts:
#	plugins/hive-plugin/ground-hive-plugin.iml

tarpdalton committed:
changed to 8082 because McAfee

tarpdalton committed:
config to correct port

tarpdalton committed:
merge master

tarpdalton committed:
refactor git parse

can fail if a commit has a parent not listed in 'latest' api call

tarpdalton committed:
cleanup

tarpdalton committed:
moved docker quickstart to wiki

tarpdalton committed:
fix for slow neo4j query

tarpdalton committed:
Merge branch 'master' of https://github.com/tarpdalton/ground

tarpdalton committed:
merge master

tarpdalton committed:
Merge remote-tracking branch 'ground-context/master'

# Conflicts:
#	.gitignore

tarpdalton committed:
Update because "Nodes." isn't needed

vsreekanti commented on ground-core/docker/config.yml:
```diff
@@ -0,0 +1,17 @@
+dbType: neo4j
```
Is it possible to not have to duplicate the config file? Or is there a reason we'd explicitly want to do this?

tarpdalton replied:
For docker to build the image it needs the config file to be in the same folder as the docker file. I can't think of a way to include the normal file.

vsreekanti replied:
Got it. We might eventually want to write a `build_docker.sh` that copies the config file and starts the container or something, but don't worry about it for now.

vsreekanti commented on ground-core/src/main/java/edu/berkeley/ground/db/Neo4jClient.java:
```diff
@@ -234,24 +235,18 @@ public Relationship getEdge(String label, List<DbDataContainer> attributes) thro
+                    "{node_id : '" + startId + "'}) WHERE a.id='" + startId +
+                    "' OR a.node_id='" + startId + "' RETURN DISTINCT e";
 
+            LOGGER.info("getDescendantEdgesByLabel query: "+query);
```
Can we remove this?

tarpdalton replied:
yes

vsreekanti commented on ground-core/src/main/java/edu/berkeley/ground/resources/NodesResource.java:
```diff
@@ -85,7 +92,9 @@ public NodeVersion createNodeVersion(@Valid NodeVersion nodeVersion, @QueryParam
         LOGGER.info("Retrieving the latest version of node " + name + ".");
-        return this.nodeFactory.getLeaves(name);
+        List<String> leaves = this.nodeFactory.getLeaves(name);
+        Set<String> set = new HashSet<>(leaves); //Dedup the leaves
```
Does the new query duplicate the leaves? If so, can (should?) we push the deduplication farther down the stack?

tarpdalton replied:
Yes that makes sense. I'll look into it

tarpdalton replied:
It looks like you fixed the problem [here](https://github.com/ground-context/ground/commit/7a62579cf7bc01d91a6cc4f14347c3639a66d7f0#diff-3b873ff5189e84e28d67a01a33b18202L92). I'll check to make sure, and then I'll remove my dedup code.

tarpdalton replied:
looks good, i removed my logic

vsreekanti commented on plugins/github-plugin/parsegitlog.py:
```diff
@@ -0,0 +1,167 @@
+from kafka import KafkaConsumer
+import configparser
+
+# TODO manage the kafka offset for restarts
```
Are all these TODOs pending? I'm assuming they can be deferred until later?

vsreekanti commented on plugins/github-plugin/parsegitlog.py:
```diff
@@ -0,0 +1,167 @@
+        print('POSTing: ' + str(json.dumps(node_version_data)))
+        r = requests.post(nodeVUrl, params=params, data=json.dumps(node_version_data), headers=headers)  # send the request
+        print(str(r) + ' Commit POSTed: ' + c['commitHash'])
+        # print(params)
```
I'm assuming these are debug prints. Can we get rid of them?

tarpdalton replied:
I'll clean this up too

tarpdalton committed:
code cleanup

cleaned up prints in python script
removed not needed dedup logic
